### PR TITLE
Support "earliest" and specific block parameters in RPC where possible

### PR DESCRIPTION
**Pull request by rphmeier**
It may not be possible whenever the database does any kind of pruning.

### CODE REVIEW

File: ethcore/src/client/mod.rs [hunk 1]
```diff
@@ -65,20 +65,48 @@ pub trait BlockChainClient : Sync + Send {
...
 	fn block_total_difficulty(&self, id: BlockID) -> Option<U256>;
 
-	/// Get address nonce.
-	fn nonce(&self, address: &Address) -> U256;
+	/// Attempt to get address nonce at given block.
+	/// May not fail on BlockID::Latest.
+	fn nonce(&self, address: &Address, id: BlockID) -> Option<U256>;
+
+	/// Get address nonce at the latest block's state.
+	fn nonce_latest(&self, address: &Address) -> U256 {
```
Comment (tomusdrw) [inline ethcore/src/client/mod.rs]: I think the most common case is to use `nonce_latest` which should stay `nonce` imho.

Can we have `nonce_at_block` or sth like this instead?

Comment (rphmeier) [inline ethcore/src/client/mod.rs]: That is how I had it initially, but I found that the breakage caused by the change in signature was actually very small in reality. `State::nonce` and `Miner::nonce` are unchanged, and this is where most code actually looks for nonces.

File: rpc/src/v1/impls/eth.rs [hunk 2]
```diff
@@ -257,6 +257,17 @@ fn pending_logs<M>(miner: &M, filter: &EthcoreFilter) -> Vec<Log> where M: Miner
 	result
 }
 
+// must be in range [-32099, -32000]
+const UNSUPPORTED_REQUEST_CODE: i64 = -32000;
+
+fn make_unsupported_err() -> Error {
+	Error {
+			code: ErrorCode::ServerError(UNSUPPORTED_REQUEST_CODE),
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth.rs]: Identation?

File: rpc/src/v1/impls/eth.rs [hunk 3]
```diff
@@ -341,27 +352,27 @@ impl<C, S, A, M, EM> Eth for EthClient<C, S, A, M, EM> where
...
 
 	fn storage_at(&self, params: Params) -> Result<Value, Error> {
 		from_params_default_third::<Address, U256>(params)
 			.and_then(|(address, position, block_number,)| match block_number {
-				BlockNumber::Pending => to_value(&U256::from(take_weak!(self.miner).storage_at(take_weak!(self.client).deref(), &address, &H256::from(position)))),
-				BlockNumber::Latest => to_value(&U256::from(take_weak!(self.client).storage_at(&address, &H256::from(position)))),
-				_ => Err(Error::invalid_params()),
+				BlockNumber::Pending => to_value(&U256::from(take_weak!(self.miner).storage_at(&*take_weak!(self.client), &address, &H256::from(position)))),
+				id => match take_weak!(self.client).storage_at(&address, &H256::from(position), id.into()) {
+						Some(s) => to_value(&U256::from(s)),
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth.rs]: Indentation??

**Pull request merged**